VultisigApp/VultisigApp/Services/Solana/SolanaService.swift

```diff
@@ -69,6 +82,135 @@ class SolanaService {
+        }
+        
+        let requestBody: [String: Any] = [
+            "jsonrpc": "2.0",
+            "id": 1,
+            "method": "getTokenAccountsByOwner",
+            "params": [
+                walletAddress,
+                ["mint": mintAddress], // SPL Token Program ID
+                ["encoding": "jsonParsed"]
+            ]
+        ]
+        
+        do {
+            let data = try await postRequest(with: requestBody)
+            let parsedData = try parseSolanaTokenResponse(jsonData: data)
+            let accounts: [SolanaTokenAccount] = parsedData.result.value
+            
+            guard let associatedAccount = accounts.first else {
+                return .empty
+            }
+            
+            // This is the address we are sending from and to
+            let pubkey = associatedAccount.pubkey
+            
+            cacheTokenAccountsByOwnerMint.set(cache_key, (data: pubkey, timestamp: Date()))
+            
+            return pubkey
+        } catch {
+            print("Error fetching tokens: \(error.localizedDescription)")
+            throw error
+        }
+    }
+    
+    func fetchTokenAccountsByOwner(for walletAddress: String) async throws -> [SolanaTokenAccount] {
+        
+        var cache_key = "getTokenAccountsByOwner_\(walletAddress)_TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"
+        if let cachedValue = await Utils.getCachedData(cacheKey: cache_key, cache: cacheTokenAccountsByOwner, timeInSeconds: CACHE_TIMEOUT_IN_SECONDS) {
+            return cachedValue
+        }
+        
+        
+        let requestBody: [String: Any] = [
+            "jsonrpc": "2.0",
+            "id": 1,
+            "method": "getTokenAccountsByOwner",
+            "params": [
+                walletAddress,
+                ["programId": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"], // SPL Token Program ID
+                ["encoding": "jsonParsed"]
+            ]
+        ]
+        
+        do {
+            let data = try await postRequest(with: requestBody)
+            let parsedData = try parseSolanaTokenResponse(jsonData: data)
+            let accounts: [SolanaTokenAccount] = parsedData.result.value
+            
+            cacheTokenAccountsByOwner.set(cache_key, (data: accounts, timestamp: Date()))
+            
+            return accounts
+        } catch {
+            print("Error fetching tokens: \(error.localizedDescription)")
+            throw error
+        }
+    }
+    
+    func fetchTokenBalance(for walletAddress: String, contractAddress: String) async throws -> String {
+        
+        do {
+            
+            let accounts: [SolanaTokenAccount] = try await fetchTokenAccountsByOwner(for: walletAddress)
+            
+            if let token = accounts.first(where: { $0.account.data.parsed.info.mint == contractAddress }) {
+                return token.account.data.parsed.info.tokenAmount.amount
+            } else {
+                return .zero
```

Comment:
Let's make the "Empty string value = nil" rule please? To not do isEmpty checks.

Reply:
Done.